BTree "the bucket being iterated changed size"

Versions:
Python 3.6
```
BTrees==4.6.0
ZODB==5.5.1
```

I have a `BTrees.OOBTree.BTree` in a ZODB which seems to have somehow become corrupted.

While our system was running we observed that the data the BTree was returning seemed suspicious (while iterating the tree, items which should have been there were not actually there). We took a snapshot of the tree and manually examined it, and determined that the structure of the BTree seems to have become corrupted. 

When we try to iterate the tree (or its keys):
```
for key in tree.keys():
    print('key: {}'.format(key))
```
We get this error:
```
Traceback (most recent call last):
  File "/home/test.py", line 40, in <module>
    for key in tree.keys():
RuntimeError: the bucket being iterated changed size
```

Interestingly, as far as I can tell our live service was not getting this exception (but it definitely was also not returning all of the data it should have been) - I'm not sure why the discrepancy between what our live service was seeing with this DB versus what I see when I download it and look through it locally.

`BTrees.check.check` shows no issues.

Obviously, I am not modifying the tree while iterating over it in my debugging environment - but when I look deeper into the structure of the tree by iterating the buckets one by one I see this:
```
Bucket: BTrees.OOBTree.OOBucket([('0231ed5f-6a2a-4675-a75a-103c6818de95', <Data object at 0x7faba5d396d8 oid 0x265b9 in <Connection at 7fabaa263cc0>>), ('03697f69-500b-4f7f-8ff1-b05c7d85b6cc', <Data object at 0x7faba5d39748 oid 0x26480 in <Connection at 7fabaa263cc0>>), ('03ac8297-38bf-4148-bd4e-df43c1e3d08b', <Data object at 0x7faba5d397b8 oid 0x262b5 in <Connection at 7fabaa263cc0>>), ('03d44486-31a2-4ed7-be24-d755f3cbff2a', <Data object at 0x7faba5d39828 oid 0x25fdb in <Connection at 7fabaa263cc0>>)])
... (a bunch more buckets that all look fine)
Bucket: BTrees.OOBTree.OOBucket([('4480e440-2d21-45f8-a909-664a85330df7', <Data object at 0x7faba5d41978 oid 0x264ad in <Connection at 7fabaa263cc0>>), ('452fc9ad-2544-4385-a034-3120a69409e8', <Data object at 0x7faba5d419e8 oid 0x264c4 in <Connection at 7fabaa263cc0>>), ('4554ce23-1ef5-4ff3-bee2-eb9233243448', <Data object at 0x7faba5d41a58 oid 0x2612f in <Connection at 7fabaa263cc0>>), ('45f92417-c80f-481a-b0d7-60110d757591', <Data object at 0x7faba5d41ac8 oid 0x25ccf in <Connection at 7fabaa263cc0>>)])
Bucket: BTrees.OOBTree.OOBucket([])
Bucket: BTrees.OOBTree.OOBucket([])
Bucket: BTrees.OOBTree.OOBucket([])
Bucket: BTrees.OOBTree.OOBucket([])
Bucket: BTrees.OOBTree.OOBucket([('91660989-9343-4f72-a076-94c8fc6c155f', <Data object at 0x7faba5ce03c8 oid 0xf769 in <Connection at 7fabaa263cc0>>), ('9176559f-e650-47bf-812b-dc4e75eaf299', <Data object at 0x7faba5ce0438 oid 0x223fa in <Connection at 7fabaa263cc0>>), ('91cd51fc-e9bd-4730-843f-b2553a06e63d', <Data object at 0x7faba5ce04a8 oid 0x22e68 in <Connection at 7fabaa263cc0>>), ('91d1b461-77ba-42b6-930b-326d7b2badbf', <Data object at 0x7faba5ce0518 oid 0x26310 in <Connection at 7fabaa263cc0>>), ('92acce7e-4e15-45c0-b954-2c25f9325731', <Data object at 0x7faba5ce0588 oid 0x221b9 in <Connection at 7fabaa263cc0>>), ('92e5b508-84a1-404b-9692-da2a7696c762', <Data object at 0x7faba5ce05f8 oid 0x22d32 in <Connection at 7fabaa263cc0>>), ... )])
... (more buckets with data here)
```

From what I can tell, 0 size buckets in the "middle" of the bucket list is not expected and is what triggers the exception I'm getting. Additionally if you look closely, it seems like a large swath of buckets is lost, because there should be a relatively even distribution of GUIDs as keys (and I see that from first digit 0-4 and first digit 9-f), but first digit 5-8 is missing.

**My questions are:**
1. Why doesn't `BTrees.check.check` notice this?
2. What could cause this type of corruption?

It's possible that this is related to #68, although they look sufficiently different to me that I figured I would post a new issue. It is also possible that this only started happening in `BTrees==4.6.0` as we moved to that version relatively recently (we were on `BTrees==4.5.1` previously) - but given how rare this is occurring we may just not have noticed it until now...

Matthew Christopher wrote at 2020-8-26 15:30 -0700:
> ...
>I took a deeper look at the difference in history between the last good history entry and the first bad one and what I observed was that we removed keys from 4 buckets between the two entries.
> ...
>The net of this is that basically we removed 5 keys from 4 buckets, with the key in bucket 159 being the last key. That bucket (oid `b'\x00\x00\x00\x00\x00\x00\r\x91'`) didn't disappear like it was (I think) supposed to.

You are right. The bucket should have been removed from the tree.

Your problem may have a stronger relation to #68
as originally assumed:
even for the good tree (which still can be iterated over)
`btree._check()` reports `AssertionError: Bucket next pointer is damaged`.
Thus, your observed problem (bucket not removed) might be
a consequence of an earlier bug (damaged bucket chain).

>The ZODB containing the BTree in question is [in this zip](https://github.com/zopefoundation/BTrees/files/5132990/db.zip).
>There are multiple BTrees in the DB: The one with the issue is: `btree = conn.root()['agent.task.history'].history`
>I've posted the code I used to gather this data [here](https://gist.github.com/matthchr/3c90b2b13c871b1407fb233e5ee2f6af).

For those who want to join the analysis, I used the following
code:
```
from pprint import pprint as pp
from ZODB.DB import DB
from ZODB.FileStorage import FileStorage
s=FileStorage("agent.db")
db=DB(s)
c=db.open()
o=c.root()['agent.task.history']
o._p_activate()
t=o.__Broken_state__["history"]
from dm.historical import *
b=c[b'\x00\x00\x00\x00\x00\x00\r\x91']
bad = getObjectAt(t, b'\x03\xda\x0bj4\xfd\x99\xee') # no longer iterable
good = getObjectAt(t, b'\x03\xda\x0bj4\xfd\x99\xe0') # still iterable
from BTrees.check import *
top=crack_btree(bad, True)
bb_p = top[2][157] # predecessor of bad bucket
bb = top[2][158] # bad bucket
good._check()
```

Dieter Maurer wrote at 2020-8-28 08:42 +0200:
>Matthew Christopher wrote at 2020-8-26 15:30 -0700:
>> ...
>Your problem may have a stronger relation to #68
>as originally assumed:
>even for the good tree (which still can be iterated over)
>`btree._check()` reports `AssertionError: Bucket next pointer is damaged`.
>Thus, your observed problem (bucket not removed) might be
>a consequence of an earlier bug (damaged bucket chain).

I have analysed this further - and indeed, the "bucket not removed"
is likely a direct consequence of the damaged bucket chain:
almost surely, a different bucket got removed (the one in the chain
at the position of the emptied bucket).

I used the following code (following code previously posted):
```
gt=crack_btree(good, True)
for i in range(len(gt[1])):
  if gt[2][i].__getstate__()[1] is not gt[2][i+1]: print(i); break
```

The result was `157`; this is the position before the emptied bucket
which was not removed (as it should have been).

Dieter Maurer wrote at 2020-8-28 09:28 +0200:
>Dieter Maurer wrote at 2020-8-28 08:42 +0200:
>>Matthew Christopher wrote at 2020-8-26 15:30 -0700:
>>> ...
>>Your problem may have a stronger relation to #68
>>as originally assumed:
>>even for the good tree (which still can be iterated over)
>>`btree._check()` reports `AssertionError: Bucket next pointer is damaged`.
>>Thus, your observed problem (bucket not removed) might be
>>a consequence of an earlier bug (damaged bucket chain).
>
>I have analysed this further - and indeed, the "bucket not removed"
>is likely a direct consequence of the damaged bucket chain:
>almost surely, a different bucket got removed (the one in the chain
>at the position of the emptied bucket).
>
>I used the following code (following code previously posted):
>```
>gt=crack_btree(good, True)
>for i in range(len(gt[1])):
>  if gt[2][i].__getstate__()[1] is not gt[2][i+1]: print(i); break
>```
>
>The result was `157`; this is the position before the emptied bucket
>which was not removed (as it should have been).

The "damaged bucket chain" was introduced by the following
transaction:
```
{'description': '',
 'obj': <BTrees.OOBTree.OOBTree object at 0x7f8b3665e158 oid 0x8 in <Connection at 7f8b366cdeb8>>,
 'size': 2961,
 'tid': b'\x03\xda\x0bi 2Wf',
 'time': DateTime('2020/08/20 01:37:7.546089 GMT+2'),
 'user_name': ''}
```

This is history record "5186" in the reversed history
obtained from `dm.historical.generateBTreeHistory(t)`.

The previous transaction (where the bucket chain is still
okay) is:
```
{'description': '',
 'obj': <BTrees.OOBTree.OOBTree object at 0x7f8b3665e1e0 oid 0x8 in <Connection at 7f8b366cdf60>>,
 'size': 2661,
 'tid': b'\x03\xda\x0bi\x02\xa9\xce\xbb',
 'time': DateTime('2020/08/20 01:37:0.624213 GMT+2'),
 'user_name': ''}
```

In the bad transaction, the bucket following the bucket with
the bad `nextbucket` pointer was emptied, the `nextbucket` pointer
of the previous bucket was correctly updated (and now points
to the bucket following the emptied bucket) **BUT**
the emptied bucket was not removed from the parent btree.


I have looked at the code: reponsible is
`BTreeTemplate.c:_BTree_set`. While the bucket chain update
is quite far away from the `data` update, I have not seen
how it could be possible to "forget" the `data` update.

I will see whether I can reproduce the behavior on my linux platform.

--- a/BTrees/_base.py
+++ b/BTrees/_base.py
@@ -1050,6 +1050,7 @@ class _Tree(_Base):
                     self._firstbucket = child._next
                     removed_first_bucket = True
             del data[index]
+            self._p_changed = True
 
         return removed_first_bucket, value
 

--- a/BTrees/tests/common.py
+++ b/BTrees/tests/common.py
@@ -91,24 +91,14 @@ class SignedMixin(object):
     KEY_RANDRANGE_ARGS = (-2000, 2001)
 
 
-class Base(SignedMixin):
-    # Tests common to all types: sets, buckets, and BTrees
+class ZODBAccess(object):
 
     db = None
-
-    def _getTargetClass(self):
-        raise NotImplementedError("subclass should return the target type")
-
-    def _makeOne(self):
-        return self._getTargetClass()()
-
-    def setUp(self):
-        super(Base, self).setUp()
-        _skip_if_pure_py_and_py_test(self)
 
     def tearDown(self):
         if self.db is not None:
             self.db.close()
+            del self.db
 
     def _getRoot(self):
         from ZODB import DB
@@ -130,6 +120,21 @@ class Base(SignedMixin):
         # "Cannot close connection joined to transaction"
         transaction.abort()
         root._p_jar.close()
+
+    
+class Base(ZODBAccess, SignedMixin):
+    # Tests common to all types: sets, buckets, and BTrees
+
+    def _getTargetClass(self):
+        raise NotImplementedError("subclass should return the target type")
+
+    def _makeOne(self):
+        return self._getTargetClass()()
+
+    def setUp(self):
+        super(Base, self).setUp()
+        _skip_if_pure_py_and_py_test(self)
+
 
     def testPersistentSubclass(self):
         # Can we subclass this and Persistent?

--- a/BTrees/tests/testPersistency.py
+++ b/BTrees/tests/testPersistency.py
@@ -1,0 +1,47 @@
+##############################################################################
+#
+# Copyright (c) 2020 Zope Foundation and Contributors.
+# All Rights Reserved.
+#
+# This software is subject to the provisions of the Zope Public License,
+# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
+# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
+# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
+# FOR A PARTICULAR PURPOSE
+#
+##############################################################################
+
+from unittest import TestCase
+
+from ..OOBTree import OOBTree, OOBTreePy
+from .common import _skip_wo_ZODB, ZODBAccess
+
+
+BUCKET_SIZE = OOBTreePy.max_leaf_size
+
+
+class TestPersistency(ZODBAccess, TestCase):
+    @_skip_wo_ZODB
+    def test_empty_bucket_persistency(self):
+        from transaction import commit
+        root = self._getRoot()
+        try:
+            # tree with 3 buckets (internal implementation details)
+            tree = OOBTree(
+                dict((i, i) for i in range(3 * BUCKET_SIZE // 2 + 2)))
+            root["tree"] = tree
+            commit()
+            # almost clear the second bucket keeping the last element
+            for i in range(BUCKET_SIZE // 2, BUCKET_SIZE - 1):
+                del tree[i]
+            commit()
+            del tree[BUCKET_SIZE - 1]  # remove the last element
+            commit()
+            tree._check()
+            tree._p_deactivate()
+            tree._check()  # fails in case of bad persistency
+        finally:
+            self._closeRoot(root)
+
+            

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -6,7 +6,7 @@
 ==================
 
 - Fix persistency bug in the Python version
-  (`#118 https://github.com/zopefoundation/BTrees/issues/118`_).
+  (`#118 <https://github.com/zopefoundation/BTrees/issues/118>`_).
 
 - Fix ``Tree.__setstate__`` to no longer accept children besides
   tree or bucket types to prevent crashes. See `PR 143

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -5,6 +5,9 @@
 4.7.3 (unreleased)
 ==================
 
+- Fix persistency bug in the Python version
+  (`#118 https://github.com/zopefoundation/BTrees/issues/118`_).
+
 - Fix ``Tree.__setstate__`` to no longer accept children besides
   tree or bucket types to prevent crashes. See `PR 143
   <https://github.com/zopefoundation/BTrees/pull/143>`_ for details.

--- a/tox.ini
+++ b/tox.ini
@@ -29,6 +29,12 @@ basepython =
 deps =
     ZODB
 
+[testenv:w_zodb-pure]
+basepython =
+    python2.7
+deps =
+    ZODB
+
 [testenv:coverage]
 basepython =
     python3.6

--- a/tox.ini
+++ b/tox.ini
@@ -3,7 +3,7 @@ envlist =
 # Jython support pending 2.7 support, due 2012-07-15 or so.  See:
 # http://fwierzbicki.blogspot.com/2012/03/adconion-to-fund-jython-27.html
 #   py27,jython,pypy,coverage,docs
-    py27,py27-pure,py35,py35-pure,py36,py37,py38,pypy,pypy3,w_zodb,coverage,docs
+    py27,py27-pure,py35,py35-pure,py36,py37,py38,pypy,pypy3,w_zodb,w_zodb-pure,coverage,docs
 
 [testenv]
 usedevelop = true

